[SPARK-13027][STREAMING] Added batch time as a parameter to updateStateByKey

aramesh117 opened the pull request:
Added RDD batch time as an input parameter to the update function in updateStateByKey.

Commit:
[SPARK-13027](STREAMING) Added batch time as a parameter to updateStateByKey.

Added RDD batch time as an input parameter to the update function in
updateStateByKey.

AmplabJenkins commented:
Can one of the admins verify this patch?

zsxwing commented:
ok to test

zsxwing commented on streaming/src/main/scala/org/apache/spark/streaming/dstream/StateDStream.scala:
```diff
@@ -42,8 +42,8 @@ class StateDStream[K: ClassTag, V: ClassTag, S: ClassTag](
-  private [this] def computeUsingPreviousRDD (
-    parentRDD: RDD[(K, V)], prevStateRDD: RDD[(K, S)]) = {
+  private [this] def computeUsingPreviousRDD (batchTime: Time,
+                                              parentRDD: RDD[(K, V)], prevStateRDD: RDD[(K, S)]) = {
```
nit: code style. Should be 
```
private [this] def computeUsingPreviousRDD (
    batchTime: Time,
    parentRDD: RDD[(K, V)], 
    prevStateRDD: RDD[(K, S)]) = {
```

aramesh117 replied:
Will fix, thank you!

SparkQA commented:
**[Test build #51011 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/51011/consoleFull)** for PR 11122 at commit [`ae6ae2d`](https://github.com/apache/spark/commit/ae6ae2db4c0481d505991203078ba43ac025d12e).

zsxwing commented on streaming/src/main/scala/org/apache/spark/streaming/dstream/StateDStream.scala:
```diff
@@ -108,7 +109,8 @@ class StateDStream[K: ClassTag, V: ClassTag, S: ClassTag](
               }
               case Some (initialStateRDD) => {
-                computeUsingPreviousRDD(parentRDD, initialStateRDD)
+                computeUsingPreviousRDD(validTime,
```
nit: could be fit into one line

aramesh117 replied:
Done.

zsxwing commented on streaming/src/main/scala/org/apache/spark/streaming/dstream/PairDStreamFunctions.scala:
```diff
@@ -453,8 +453,14 @@ class PairDStreamFunctions[K, V](self: DStream[(K, V)])
       rememberPartitioner: Boolean
-    ): DStream[(K, S)] = ssc.withScope {
-     new StateDStream(self, ssc.sc.clean(updateFunc), partitioner, rememberPartitioner, None)
+    ): DStream[(K, S)] = {
```
nit: could you remove the outer `{}` like other methods?

aramesh117 replied:
Done.

zsxwing commented on streaming/src/main/scala/org/apache/spark/streaming/dstream/PairDStreamFunctions.scala:
```diff
@@ -499,9 +505,35 @@ class PairDStreamFunctions[K, V](self: DStream[(K, V)])
-    ): DStream[(K, S)] = ssc.withScope {
-     new StateDStream(self, ssc.sc.clean(updateFunc), partitioner,
-       rememberPartitioner, Some(initialRDD))
+    ): DStream[(K, S)] = {
```
nit: could you remove the outer `{}` like other methods?

aramesh117 replied:
Done.

zsxwing commented:
Could you add a method with `rememberPartitioner: Boolean`?

aramesh117 commented:
@zsxwing Sure.

zsxwing commented on streaming/src/test/scala/org/apache/spark/streaming/BasicOperationsSuite.scala:
```diff
@@ -467,6 +467,73 @@ class BasicOperationsSuite extends TestSuiteBase {
+        }
+      }
+      val x = s.map(x => (x, 1))
+        x.updateStateByKey[StreamingState](updateFunc = updateFunc,
```
nit: not need to create `x`. Just `s.map(x => (x, 1)).updateStateByKey`

aramesh117 replied:
Sorry yes, I had created a variable to find its type while debugging an issue. I'll restore the original.

SparkQA commented:
**[Test build #51011 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/51011/consoleFull)** for PR 11122 at commit [`ae6ae2d`](https://github.com/apache/spark/commit/ae6ae2db4c0481d505991203078ba43ac025d12e).
 * This patch passes all tests.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test PASSed.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/51011/
Test PASSed.

Commit:
[SPARK-13027][STREAMING] Addressing comments.

- Fixed code style issues.
- Fixed inconsistent ssc.withScope invocation.
- Added DStream closure cleaning test for new update function.

SparkQA commented:
**[Test build #51019 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/51019/consoleFull)** for PR 11122 at commit [`d69f46b`](https://github.com/apache/spark/commit/d69f46b14fab3a852e05761a7cc0fb99c535fa7a).

SparkQA commented:
**[Test build #51019 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/51019/consoleFull)** for PR 11122 at commit [`d69f46b`](https://github.com/apache/spark/commit/d69f46b14fab3a852e05761a7cc0fb99c535fa7a).
 * This patch passes all tests.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test PASSed.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/51019/
Test PASSed.

aramesh117 commented:
@zsxwing Any more issues that you can see?

zsxwing commented on streaming/src/main/scala/org/apache/spark/streaming/dstream/PairDStreamFunctions.scala:
```diff
@@ -452,11 +452,15 @@ class PairDStreamFunctions[K, V](self: DStream[(K, V)])
+    new StateDStream(self, newUpdateFunc, partitioner, rememberPartitioner, None)
   }
 
+
```
nit: extra line

aramesh117 replied:
Done.

zsxwing commented on streaming/src/main/scala/org/apache/spark/streaming/dstream/PairDStreamFunctions.scala:
```diff
@@ -498,10 +502,36 @@ class PairDStreamFunctions[K, V](self: DStream[(K, V)])
+    * @tparam S State type
+    */
+  def updateStateByKey[S: ClassTag]
+  (updateFunc: (Time, K, Seq[V], Option[S]) => Option[S],
```
nit: move `(` to the above line.

zsxwing commented on streaming/src/main/scala/org/apache/spark/streaming/dstream/PairDStreamFunctions.scala:
```diff
@@ -498,10 +502,36 @@ class PairDStreamFunctions[K, V](self: DStream[(K, V)])
+    */
+  def updateStateByKey[S: ClassTag]
+  (updateFunc: (Time, K, Seq[V], Option[S]) => Option[S],
+   partitioner: Partitioner,
```
Please use 4 spaces indention for the parameters

zsxwing commented on streaming/src/main/scala/org/apache/spark/streaming/dstream/PairDStreamFunctions.scala:
```diff
@@ -498,10 +502,36 @@ class PairDStreamFunctions[K, V](self: DStream[(K, V)])
+    val newUpdateFunc = (time: Time, iterator: Iterator[(K, Seq[V], Option[S])]) => {
+      iterator.flatMap(t => cleanedFunc(time, t._1, t._2, t._3).map(s => (t._1, s)))
+    }
+    new StateDStream(self, newUpdateFunc, partitioner,
```
nit: can be fit into one line

zsxwing commented on streaming/src/main/scala/org/apache/spark/streaming/dstream/PairDStreamFunctions.scala:
```diff
@@ -498,10 +502,36 @@ class PairDStreamFunctions[K, V](self: DStream[(K, V)])
+      cleanedFunc(it)
+    }
+    new StateDStream(self, newUpdateFunc, partitioner,
+      rememberPartitioner, Some(initialRDD))
```
nit: can be just one line

zsxwing commented on streaming/src/test/scala/org/apache/spark/streaming/DStreamClosureSuite.scala:
```diff
@@ -171,6 +171,10 @@ class DStreamClosureSuite extends SparkFunSuite with BeforeAndAfterAll {
     val updateF1 = (_: Seq[Int], _: Option[Int]) => { return; Some(1) }
     val updateF2 = (_: Iterator[(Int, Seq[Int], Option[Int])]) => { return; Seq((1, 1)).toIterator }
+    val updateF3 = (_: Time, _: Int, _: Seq[Int], _: Option[Int]) => {
+      return
```
return?

zsxwing commented:
Just some style issues. Almost LGTM. I would like to let @tdas take a final look as this is adding new API.

Commit:
[Streaming] Addressing style comments.

SparkQA commented:
**[Test build #52888 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/52888/consoleFull)** for PR 11122 at commit [`17cdabd`](https://github.com/apache/spark/commit/17cdabdf75a759675bf7defb635017a384d1c880).

aramesh117 commented:
I've fixed your comments. Sorry for the delay.

SparkQA commented:
**[Test build #52888 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/52888/consoleFull)** for PR 11122 at commit [`17cdabd`](https://github.com/apache/spark/commit/17cdabdf75a759675bf7defb635017a384d1c880).
 * This patch passes all tests.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test PASSed.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/52888/
Test PASSed.

aramesh117 commented:
@tdas Have you had a chance to look at the pull request?

aramesh117 commented:
@zsxwing Seems like @tdas is busy. Can you merge this if possible before Spark 2.0?

zzcclp commented:
@aramesh117 @zsxwing @tdas Is there any process at this pr?

zzcclp commented:
ping @zsxwing @tdas

aramesh117 commented:
@zzcclp This pr was completed and approved by @zsxwing but hasn't been merged for some reason. I've tried to reach @tdas but he hasn't responded. What should we do? Now, there appear to be merge conflicts.

zzcclp commented:
@aramesh117 you should update code based on master branch to resovlve conflicts, and then, jenkins will  test this pr automatically.

zsxwing commented:
@aramesh117 sorry that I forgot this PR. Could you update to resolve conflicts, please? Thanks!

zzcclp commented:
ping @aramesh117

aramesh117 commented:
@zzcclp and @zsxwing I will fix conflicts now.

Commit:
Fixing scaladoc style issue.

aramesh117 commented:
@zsxwing and @zzcclp please take a look when you can - I've fixed all conflicts. I've run all tests and verified that all tests relevant to my change have passed. I would greatly appreciate it if we could merge it in soon. Thank you very much.

SparkQA commented:
**[Test build #68501 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/68501/consoleFull)** for PR 11122 at commit [`fe68b6c`](https://github.com/apache/spark/commit/fe68b6c03300a37799ccaad2ee554bde005c8f6f).

SparkQA commented:
**[Test build #68501 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/68501/consoleFull)** for PR 11122 at commit [`fe68b6c`](https://github.com/apache/spark/commit/fe68b6c03300a37799ccaad2ee554bde005c8f6f).
 * This patch passes all tests.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test PASSed.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/68501/
Test PASSed.

zzcclp commented:
It sounds good. @zsxwing could you take a look at this pr? does this pr plan to be merged into branch-2.1?

zsxwing commented:
retest this please.

LGTM. Let's run the test again. Will merge after test passes.

SparkQA commented:
**[Test build #3426 has started](https://amplab.cs.berkeley.edu/jenkins/job/NewSparkPullRequestBuilder/3426/consoleFull)** for PR 11122 at commit [`fe68b6c`](https://github.com/apache/spark/commit/fe68b6c03300a37799ccaad2ee554bde005c8f6f).

SparkQA commented:
**[Test build #3426 has finished](https://amplab.cs.berkeley.edu/jenkins/job/NewSparkPullRequestBuilder/3426/consoleFull)** for PR 11122 at commit [`fe68b6c`](https://github.com/apache/spark/commit/fe68b6c03300a37799ccaad2ee554bde005c8f6f).
 * This patch passes all tests.
 * This patch merges cleanly.
 * This patch adds no public classes.

zsxwing commented:
Merging to master and 2.1. @aramesh117 apologize for the delay again. Thanks a lot for your patience.

aramesh117 commented:
@zsxwing and @zzcclp thank you so much. This is much appreciated. :)
